File: my-library/Pollard.cpp
// unchecked

#include <iostream>
#include <cmath>
#include <random>

using namespace std;

#define int long long

int sum(int x, int y, int mod) { return x + y < mod ? x + y : x + y - mod; }
int mul(int x, int y, int mod) {
	int res = 0;
	while (y) {
		if (y & 1) {
			res = sum(res, x, mod);
		}
		x = sum(x, x, mod);
		y >>= 1;
	}
	return res;
}

mt19937 rnd(239);

int gcd(int a, int b) { return b ? gcd(b, a % b) : a; }

int pollard(int n) {
	if (n % 2 == 0) {
		return 2;
	}
	else if ((int)sqrt(n) * (int)sqrt(n) == n) {
		return sqrt(n);
	}
	else {
		auto nxt = [&](int seed) { return sum(mul(seed, seed, n), 3, n); };
		int x = rnd() % n;
		int y = x;
		for (int it = 0; it < 1e6; ++it) {
			if (x != y & gcd(abs(x - y), n) != 1) {
				return gcd(abs(x - y), n);
			}
			x = nxt(x);
			y = nxt(nxt(y));
		}
		return -1;
	}
}

File: my-library/Math/Number Theory/Pollard.py
# unchecked

import random

def gcd(a, b):
    if b == 0: return a
    else: return gcd(b, a % b)


def nxt(seed):
    return seed * seed + 3


def pollard(n):
    if n % 2 == 0: return 2
    elif int(n ** 0.5) ** 2 == n: return int(n ** 0.5)
    else:
        x = y = random.randint(0, n - 1)
        for it in range(1000000):
            if x != y and gcd(abs(x - y), n) != 1:
                return gcd(abs(x - y), n)
            x = nxt(x) % n
            y = nxt(nxt(y)) % n
        return -1
